### DISCUSSION TITLE
stroke inside 、stroke outside、stroke center

### DISCUSSION THREAD

**Discussion by zhe-he** [fabricjs/fabric.js#9727] [Show and tell]
<img width="454" alt="image" src="https://github.com/fabricjs/fabric.js/assets/11306583/bdf1f2d2-d74a-47ca-b876-a6565dfcffce">

[simple demo](https://codesandbox.io/p/devbox/fabric-stroke-side-x6s4pz?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clptda5ls00073b6hub7q0rb8%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clptda5lo00023b6h7o2lx4gg%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clptda5lr00043b6h1ko80kdv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clptda5ls00063b6hyx3tfc52%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clptda5lo00023b6h7o2lx4gg%2522%253A%257B%2522id%2522%253A%2522clptda5lo00023b6h7o2lx4gg%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clptda5ls00063b6hyx3tfc52%2522%253A%257B%2522id%2522%253A%2522clptda5ls00063b6hyx3tfc52%2522%252C%2522activeTabId%2522%253A%2522clto50vgc00ea3b6gqczq5drm%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clto50vgc00ea3b6gqczq5drm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clptda5lr00043b6h1ko80kdv%2522%253A%257B%2522id%2522%253A%2522clptda5lr00043b6h1ko80kdv%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clptda5lr00033b6hp9wzt2bi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clptda5lr00033b6hp9wzt2bi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

[Complete demo](https://codesandbox.io/p/devbox/fabric-stroke-patch-nnyrlw?file=%2Fsrc%2Ftest-stroke-side.tsx%3A29%2C17)

**Comment by ShaMan123** [comment]
very cool and super useful feature

**Comment by vshepel** [comment]
@zhe-he Thanks for your snippet, very helpful for me. Also, do you know how to keep the original width/height of the rectangle? Because right now, if a 100x100px rectangle has a 10px stroke, it is drawn 110x110px

**Comment by zhe-he** [comment]
@vshepel 
Here is the complete version.
[Complete demo](https://codesandbox.io/p/devbox/fabric-stroke-patch-nnyrlw?file=%2Fsrc%2Ftest-stroke-side.tsx%3A29%2C17)

**Comment by zhe-he** [comment]
Fabric currently defaults to a center `stroke side` for outlines which is not configurable, and the calculation of an object's bounds includes the `stroke width`. To support different types of `stroke side`, it would be necessary to override Fabric's method for calculating an object's bounds, such that the bounds do not include the `stroke width`. Currently, Fabric draws strokes and fills simultaneously; to support different `stroke side` types, they would need to be drawn separately. The complete example includes detailed code.

**Comment by ixover** [comment]
Perfect, Super useful feature thank you

**Comment by mauricekindermann** [comment]
Cool idea, I just ran into this same problem with clipPath. If only it could trace any shape's border, and didn't distore on resize. Like @zhe-he said you'd really want to bake this into fabric to work as a naive feature, so you could set something like a strokePosition.
